src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,936 @@
+/*-------------------------------------------------------------------------
+ *
+ * alter_table.c
+ *	  Routines relation to the altering of tables.
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "fmgr.h"
+
+#include "access/hash.h"
+#include "catalog/dependency.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/coordinator_protocol.h"
+#include "distributed/distribution_column.h"
+#include "distributed/listutils.h"
+#include "distributed/metadata/dependency.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_logical_planner.h"
+#include "distributed/multi_partitioning_utils.h"
+#include "distributed/worker_protocol.h"
+#include "executor/spi.h"
+#include "nodes/pg_list.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+
+
+/* Table Conversion Types */
+#define UNDISTRIBUTE_TABLE 'u'
+#define ALTER_DISTRIBUTED_TABLE 'a'
+#define ALTER_TABLE_SET_ACCESS_METHOD 'm'
+
+struct TableConversion;
+
+typedef void (*TableConversionFunction)(struct TableConversion *);
+
+typedef struct TableConversion
+{
+	/*
+	 * Determines type of conversion: UNDISTRIBUTE_TABLE,
+	 * ALTER_DISTRIBUTED_TABLE, ALTER_TABLE_SET_ACCESS_METHOD.
+	 *
+	 *
+	 */
+	char conversionType;
+
+	/* Oid of the table to do conversion on */
+	Oid relationId;
+
+	/*
+	 * Options to do conversions on the table
+	 * distributionColumn is the name of the new distribution column,
+	 * shardCountIsNull is if the shardCount variable is not given
+	 * shardCount is the new shard count,
+	 * colocateWith is the name of the table to colocate with, 'none', or
+	 * 'default'
+	 * accessMethod is the name of the new accessMethod for the table
+	 */
+	char *distributionColumn;
+	bool shardCountIsNull;
+	int shardCount;
+	char *colocateWith;
+	char *accessMethod;
+
+	/*
+	 * cascadeToColocatedIsNull is if the cascateToColocated
+	 * variable is given
+	 * cascateToColocated determines whether to cascated
+	 * shardCount and colocateWith will be cascaded to the
+	 * currently colocated tables
+	 */
+	bool cascadeToColocatedIsNull;
+	bool cascadeToColocated;
```

Comment:
could this be an enum? (something like CASCADE_TO_COLOCATED_NO, CASCADE_TO_COLOCATED_YES, CASCADE_TO_COLOCATED_UNSPECIFIED)